### ISSUE TITLE
Cache decoded scene images: 6.1 s of every run is PNG decode

### ISSUE THREAD

**Issue by nandometzger** [nandometzger/metal-gauss#6]
Every process pays **~8.4 s before the first training step**, and it is almost all Metal shader compilation.

`bench/results/startup_cold.json` and `startup_warm.json`:

| | startup_s |
|---|---|
| cold, extension rebuilt | 8.4, 8.7, 8.6 |
| warm | 8.7, 8.3, 8.4, 8.3 |

**Cold and warm are identical.** The C++/ObjC++ extension is cached to a `.so` and paid once ever, so it does not appear here. What is left is `newLibraryWithSource`, which compiles the shaders from source in every process and which torch cannot cache.

## Why it matters

This is the whole reason the README's own caveat reads the way it does:

> msplat is 1.3–1.8x faster per step and owns every budget under ~0.3 min. Our floor there is ~8.4 s of fixed startup.

At the fast end of the Pareto front, startup *is* the number. A 100-iteration run costs about 0.22 min, of which the training is a few seconds. Removing even half of the 8.4 s would move the crossover point where msplat stops winning.

## What to try

Emit a compiled `.metallib` at build time and load it with `newLibraryWithData` instead of compiling from source at import. Invalidation has to key on the shader source, so a stale library is never silently used.

Measure with `python bench/startup_profile.py`, which already separates the extension build, shader compilation, and the filesystem page cache. Startup is `harness_wall_s - train_only_wall_s`, since the trainer's own timer begins after scene loading and kernel setup.

Anyone taking this on should read `CONTRIBUTING.md` on timing first: Apple Silicon runs short bursts at boost clock, so a first measurement is optimistic.

**Comment by nandometzger** [comment]
## The 8.4 s is not shader compilation

I profiled this before starting on the `.metallib` work, and the measurement does not support the diagnosis. Shader compilation is **~30 ms**, about 0.4% of startup.

Isolating the three calls inside `_load()` in a fresh process (13 kernels, 1045 lines of MSL across the five `.metal` files):

```
cpp_extension load    0.05 s
newLibraryWithSource  0.03 s   <- the call this issue proposes to cache
first kernel + PSOs   0.13 s
```

Then timing every phase from process start to the trainer's `t0` (`train.py:274`), on `lego` at `max_resolution=800`:

| phase | time | share |
|---|---|---|
| `load_blender` (PNG decode + LANCZOS resize) | **6.12 s** | **70%** |
| `import torch` + numpy | 0.77 s | 9% |
| `make_optimizer` | 0.69 s | 8% |
| `init_params` | 0.56 s | 6% |
| `_load` (extension + shaders, all in) | 0.55 s | 6% |
| **total to `t0`** | **8.70 s** | |

That total matches what this issue reports, and `bench/startup_profile.py` reproduced the published figure unchanged on the same machine (8.60 s / 8.80 s warm, 13.10 s on the first run). So the 8.4 s is real; only its attribution was wrong.

## How the inference went wrong

The issue's argument is that cold and warm are identical, the `.so` is cached and paid once, therefore the remainder must be `newLibraryWithSource`. But `startup_s` is `harness_wall_s - train_only_wall_s`, which spans an entire subprocess: interpreter start, torch import, scene loading, parameter init, optimizer construction. Cold ≡ warm is equally consistent with decoding 100 800x800 PNGs on every run — 41 MB, page-cache-warm in both arms, so it costs the same either way.

This is the harness-attribution pattern that `bench/runner.py` already documents, in a new place: the number was right, the label on it was not.

## What the lever actually is

`metal_gauss/blender.py:49-53` opens and LANCZOS-resizes each view with PIL, 100 views per run, with nothing cached between runs. A decoded-tensor cache keyed on scene and `max_resolution` addresses ~70% of startup instead of 0.4%.

That is also what moves the README's Pareto caveat. The "~8.4 s of fixed startup" that msplat wins against is mostly a PNG decoder, not the Metal path.

Two smaller notes on the original plan:

- `xcrun metal` and `xcrun metallib` are unavailable under Command Line Tools, so the build-time `.metallib` step could not run on the machine this project develops on. It would have needed a full-Xcode CI box and a shipped artifact.
- Now that #12 and #13 have landed, `initLibrary` pins MSL 3.0 explicitly, so any future library-caching work has to key invalidation on the language version as well as the source hash.

## Reproducing

Environment: M5 MacBook Air, macOS 26.6.2, Python 3.12, PyTorch 2.14.0, `main` at e820561.

`bench/startup_profile.py` gives the total. The phase split came from timing `load_blender`, `init_params`, `make_optimizer` and `_load` individually, and the shader number from calling `ext.init(src)` on its own with the extension already built. Happy to send the scripts if useful, though both are a few lines around `time.perf_counter()`.

## Suggestion

Retitle this to the image decode cache, or close it and open a fresh one, and keep the `.metallib` idea only as a note that it is worth ~30 ms.
